lms/djangoapps/certificates/badge_handler.py

```diff
@@ -0,0 +1,151 @@
+            )
+            raise
+
+    def get_headers(self):
+        """
+        Headers to send along with the request-- used for authentication.
+        """
+        return {'Authorization': 'Token {}'.format(settings.BADGR_API_TOKEN)}
+
+    def ensure_badge_created(self, mode):
+        """
+        Verify a badge has been created for this mode of the course, and, if not, create it
+        """
+        if self.course_slug(mode) in BadgeHandler.badges:
+            return
+        response = requests.get(self.badge_url(mode), headers=self.get_headers())
+        if response.status_code != 200:
+            self.create_badge(mode)
+        BadgeHandler.badges[self.course_slug(mode)] = True
+
+    def create_badge(self, mode):
+        """
+        Create the badge spec for a course's mode.
+        """
+        course = modulestore().get_course(self.course_key)
+        if mode not in settings.BADGR_IMAGE_SOURCES:
+            raise KeyError("No image specified for course mode {} in BADGR_IMAGE_SOURCES!".format(mode))
+        image_name = settings.BADGR_IMAGE_SOURCES[mode]
+        image = open(image_name, 'rb')
+        # We don't want to bother validating the file any further than making sure we can detect its MIME type,
+        # for HTTP. The Badgr-Server should tell us if there's anything in particular wrong with it.
+        content_type, __ = mimetypes.guess_type(image_name)
+        if not content_type:
+            raise ValueError(
+                "Could not determine content-type of image! Make sure it is a properly named .png or .svg file."
+            )
+        files = {'image': (image_name, image, content_type)}
+        try:
+            domain = unicode(Site.objects.get_current().domain)
+        except Site.DoesNotExist:
+            domain = u'example.com'
+        about_path = reverse('about_course', kwargs={'course_id': unicode(self.course_key)})
+        data = {
+            'name': course.display_name,
+            'criteria': u'http://{}{}'.format(domain, about_path),
+            'slug': self.course_slug(mode)
+        }
+        result = requests.post(self.badge_create_url, headers=self.get_headers(), data=data, files=files)
+        self.log_if_raised(result)
+
+    def create_assertion(self, user, mode):
+        """
+        Register an assertion with the Badgr server for a particular user in a particular course mode for
+        this course.
+        """
+        response = requests.post(self.assertion_url(mode), headers=self.get_headers(), data={'email': user.email})
+        self.log_if_raised(response)
+        assertion, __ = BadgeAssertion.objects.get_or_create(course_id=self.course_key, user=user)
+        assertion.data = response.json()
+        assertion.save()
+        tracker.emit(
+            'edx.badges.assertion.created', {
+                'user': user.id,
+                'course_id': unicode(self.course_key),
+                'mode': mode,
+                'assertion': assertion.data
```

Comment:
@Kelketek is right that it will vary a little, but not much. There is an optional "evidence" field that may appear in some badges. Otherwise the set of fields is fairly constant. BadgeClasses have more variance with more optional fields and an unconstrained-length description field (though issuers are encouraged to keep their descriptions short)